driftbase/api/messages.py

```diff
@@ -95,28 +95,33 @@ def fetch_messages(exchange, exchange_id, min_message_number=0, rows=None):
         i += 1
         if curr_message_number < min_message_number:
             break
+        highest_processed_message_number = max(curr_message_number, highest_processed_message_number)
         expires = datetime.datetime.fromisoformat(message["expires"][:-1]) # remove trailing 'Z'
         if expires > utcnow():
             messages.append(message)
-            log.info("Message %s ('%s') has been retrieved from queue '%s' in "
+            log.debug("Message %s ('%s') has been retrieved from queue '%s' in "
                      "exchange '%s-%s' by player %s",
                      message["message_number"], message["message_id"],
                      message["queue"], exchange, exchange_id, my_player_id)
 
             if rows and len(messages) >= rows:
                 break
         else:
-            log.info("Expired message %s ('%s') was removed from queue '%s' in "
+            log.debug("Expired message %s ('%s') was removed from queue '%s' in "
                      "exchange '%s-%s' by player %s",
                      message["message_number"], message["message_id"],
                      message["queue"], exchange, exchange_id, my_player_id)
 
+    # If there were only expired messages, make sure we skip those next time
+    if len(messages) == 0 and highest_processed_message_number > 0:
+        g.redis.conn.setex(redis_seen_key, MESSAGE_EXCHANGE_TTL, highest_processed_message_number)
+
     messages.sort(key=operator.itemgetter("message_number"), reverse=True)
     rows = rows or len(messages)
     ret = collections.defaultdict(list)
     for m in messages[:rows]:
         ret[m["queue"]].append(m)
-    return ret
+    return ret, highest_processed_message_number
```

Comment:
Is anyone looking at the returned highest_processed_message_number ?

Reply:
The idea is to send that back to clients so they can use that as an anchor for the next poll, but it's not used yet

Reply:
Right now if messages are discarded, clients won't know that they should be asking for messages *after* those IDs, since they never get them